File: 1027-longest-arithmetic-subsequence/1027-longest-arithmetic-subsequence.java
class Solution {
    int[][] dp = new int[1001][1003];
    
    public int find(int ind, int[] nums, int diff) {
        if(dp[ind][diff+501] != -1) return dp[ind][diff+501];
        
        int a = 1;
        
        if(diff == -501) {
            for(int i=ind+1; i<nums.length; ++i)
                a = Math.max(a, Math.max(1+find(i, nums, nums[i]-nums[ind]), find(i, nums, diff)));
        } else {
            for(int i=ind+1; i<nums.length; ++i) {
                if(nums[i]-nums[ind] == diff)
                    a = Math.max(a, 1+find(i, nums, diff));
            }
        }
        
        return dp[ind][diff+501] = a;
    }
    
    public int longestArithSeqLength(int[] nums) {
        Arrays.stream(dp).forEach(a -> Arrays.fill(a, -1));
        return find(0, nums, -501);
    }
}

File: 1027-longest-arithmetic-subsequence/1027-longest-arithmetic-subsequence.cpp
class Solution {
public:
    int longestArithSeqLength(vector<int>& nums) {
        int n = nums.size();
        vector<vector<int>> dp(n,vector<int>(2000,0));
        int ans=0;
        for(int i=0;i<n;++i)
        {
            for(int j=i+1;j<n;++j)
            {
                int d=nums[j]-nums[i]+1000;
                dp[j][d]=dp[i][d]>0?1+dp[i][d]:2;
                ans=max(ans,dp[j][d]);
            }
        }
        return ans;
    }
};